examples/examples/rpc_middleware_rate_limiting.rs

```diff
@@ -0,0 +1,161 @@
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Example middleware to rate limit based on the number
+//! JSON-RPC calls.
+//!
+//! As demonstrated in this example any state must be
+//! stored in something to provide interior mutability
+//! such as `Arc<Mutex>`
+
+use jsonrpsee::core::{async_trait, client::ClientT};
+use jsonrpsee::server::middleware::{Context, RpcServiceBuilder, RpcServiceT};
+use jsonrpsee::server::Server;
+use jsonrpsee::types::{ErrorObject, Request};
+use jsonrpsee::ws_client::WsClientBuilder;
+use jsonrpsee::{rpc_params, MethodResponse, RpcModule};
+use std::net::SocketAddr;
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+#[derive(Debug, Copy, Clone)]
+struct Rate {
+	num: u64,
+	period: Duration,
+}
+
+#[derive(Debug, Copy, Clone)]
+enum State {
+	Deny { until: Instant },
+	Allow { until: Instant, rem: u64 },
+}
+
+#[derive(Clone)]
+pub struct RateLimit<S> {
+	service: S,
+	state: Arc<Mutex<State>>,
+	rate: Rate,
+}
+
+impl<S> RateLimit<S> {
+	fn new(service: S, rate: Rate) -> Self {
+		let period = rate.period;
+		let num = rate.num;
+
+		Self {
+			service,
+			rate,
+			state: Arc::new(Mutex::new(State::Allow { until: Instant::now() + period, rem: num + 1 })),
+		}
+	}
+}
+
+#[async_trait]
+impl<'a, S> RpcServiceT<'a> for RateLimit<S>
+where
+	S: Send + Sync + RpcServiceT<'a>,
+{
+	async fn call(&self, req: Request<'a>, meta: &Context) -> MethodResponse {
+		let now = Instant::now();
+
+		let is_denied = {
+			let mut lock = self.state.lock().unwrap();
+			let next_state = match *lock {
+				State::Deny { until } => {
+					if now > until {
+						State::Allow { until: now + self.rate.period, rem: self.rate.num - 1 }
+					} else {
+						State::Deny { until }
+					}
+				}
+				State::Allow { until, rem } => {
+					if now > until {
+						State::Allow { until: now + self.rate.period, rem: self.rate.num - 1 }
+					} else {
+						let n = rem - 1;
+						if n > 0 {
+							State::Allow { until: now + self.rate.period, rem: n }
+						} else {
+							State::Deny { until }
+						}
+					}
+				}
+			};
+
+			*lock = next_state;
+			matches!(next_state, State::Deny { .. })
+		};
+
+		if is_denied {
+			MethodResponse::error(req.id, ErrorObject::borrowed(-32000, "RPC rate limit", None))
```

Comment:
For a production level rate limiter, we need the ability to terminate connection and reject new connection from specific IP / header. Is this possible?

Reply:
Yes, we have introduced a this context struct https://github.com/paritytech/jsonrpsee/pull/1215/files#diff-68ed87005f3d8ed3f1b846a7db86e3f0ae76177f84ee16f9def28fe0f4fa1ea9R74-R87.

So it should be possible and based on how you instantiate the "middleware layer closure" you pick whether to create a new state per connection connection or globally on the server.

For example:

```rust
let global_cnt = Counter::default();
let rpc_middleware = RpcServiceBuilder::new()
		// This state is created per connection.
		.layer_fn(|service| CallsPerConn { service, count: Counter::default() })
		// This state is shared by all connections.
		.layer_fn(move |service| GlobalCalls { service, count: global_cnt.clone() });
```

Reply:
But it's nothing in here to be able to terminate the connection in here just allow/deny a specific RPC but I reckon that we need another middleware for that

Reply:
Yeah I want to drop the connection with a peer and prevent new connection from the same IP. Let me know if you want me to write an issue about this.

Reply:
Yeah, I get that.

I'll write up on issue on it would most likely be a follow-up this PR and we would need something to apply on the HTTP level to ban certain IP addresses or other stuff already at the HTTP level but I'm afraid the we may need three different APIs for it.